Adding hostgroup provider and type

flyinbutrs opened the pull request:
### Hostgroup Provider and Type
This pull request includes a provider and type for foreman hostgroups based on the existing foreman_smartproxy provider and type.  

It supports:
* Creating and Destroying hostgroups
* Assigning Smartproxy
* Assigning Puppetclasses
* Assigning default environment

### oauth facts
* foreman_ouath_consumer_key
* foreman_ouath_consumer_secret

This one I'm curious if you will want or not, but I created facts that pull the oauth key and secret out of /etc/foreman/settings.yaml. I realize this may be considered a security risk, so I'm curious to hear other people's thoughts on this.  We use a lot of automated foreman deployments which manage themselves through manifests, and these facts seemed required to allow that sort of self-management.

flyinbutrs commented:
FYI, 99% sure this would pass travis if it were rerun, the only two failures were timeouts, not actual errors.

mmoll commented:
yes, these particular two combinations are known to time out at the moment, sorry for that

ekohl commented:
Regarding the security: at first I thought that if you have the permission to read those files, then you can read them anyway. Then again, facts are pushed to the foreman and that may allow users who should have no permission to read them. I think we can merge the hostgroup provider + type without a problem so I'd prefer to split it into 2 PRs.

flyinbutrs commented:
After some more thought, I agree the secret / key shouldn't be stored as facts.  I've removed the facts and instead setup the key/secret to be read in by the type from /etc/foreman/settings.yaml if no values are provided.  That way they are still usable for local management without storing them as facts.

ekohl commented:
I think the code looks good and makes sense, but not an expert on this part.

flyinbutrs commented:
ok @ekohl who should I direct this towards? @mmoll?

mmoll commented:
Yeah, I'll test this. There are some codestyle "problems" I've seen when running rubocop over the files, but I don't think it's making much sense to poke these until we enfore style also for Ruby code in Puppet modules. After that @domcleal and/or @mbacovsky should have a look regarding API/apipie usage and merge, if there are no further comments.

flyinbutrs commented:
How would I run that test myself? My linter didn't show any issues, but I'm all in favor of keeping to style guidelines.

mmoll commented:
```
mmoll@marduk:/tmp/puppet-foreman$ rubocop  lib/puppet/provider/foreman_hostgroup/rest_v2.rb|less
Inspecting 1 file
W

Offenses:

lib/puppet/provider/foreman_hostgroup/rest_v2.rb:2:1: C: Extra empty line detected at block body beginning.
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:3:11: C: Use the new Ruby 1.9 hash syntax.
  confine :feature => :apipie_bindings
          ^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:6:81: C: Line is too long. [84/80]
    # body = JSON.parse(e.response)["error"]["full_messages"].join(" ") rescue 'N/A'
                                                                                ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:7:81: C: Line is too long. [95/80]
    fail "Hostgroup #{resource[:name]} cannot be registered: (#{e.message}): #{e.backtrace[0]}"
                                                                                ^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:15:3: C: Assignment Branch Condition size for api is too high. [27.95/15]
  def api
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:15:3: C: Method has too many lines. [31/10]
  def api
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:22:9: C: Use raise instead of fail to rethrow exceptions.
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Key available"
        ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:22:81: C: Line is too long. [97/80]
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Key available"
                                                                                ^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:30:81: C: Line is too long. [85/80]
        secret = YAML.load_file('/etc/foreman/settings.yaml')[:oauth_consumer_secret]
                                                                                ^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:32:9: C: Use raise instead of fail to rethrow exceptions.
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Secret available"
        ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:32:81: C: Line is too long. [100/80]
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Secret available"
                                                                                ^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:36:38: C: Redundant curly braces around a hash parameter.
    @api ||= ApipieBindings::API.new({
                                     ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:37:7: C: Use the new Ruby 1.9 hash syntax.
      :uri => resource[:base_url],
      ^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:37:7: C: Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.
      :uri => resource[:base_url],
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:38:7: C: Use the new Ruby 1.9 hash syntax.
      :api_version => 2,
      ^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:39:7: C: Use the new Ruby 1.9 hash syntax.
      :oauth => {
      ^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:40:9: C: Use the new Ruby 1.9 hash syntax.
        :consumer_key    => key,
        ^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:41:9: C: Use the new Ruby 1.9 hash syntax.
        :consumer_secret => secret
        ^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:43:7: C: Use the new Ruby 1.9 hash syntax.
mmoll@marduk:/tmp/puppet-foreman$ rubocop  lib/puppet/provider/foreman_hostgroup/rest_v2.rb|less -R
Inspecting 1 file
W

Offenses:

lib/puppet/provider/foreman_hostgroup/rest_v2.rb:2:1: C: Extra empty line detected at block body beginning.
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:3:11: C: Use the new Ruby 1.9 hash syntax.
  confine :feature => :apipie_bindings
          ^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:6:81: C: Line is too long. [84/80]
    # body = JSON.parse(e.response)["error"]["full_messages"].join(" ") rescue 'N/A'
                                                                                ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:7:81: C: Line is too long. [95/80]
    fail "Hostgroup #{resource[:name]} cannot be registered: (#{e.message}): #{e.backtrace[0]}"
                                                                                ^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:15:3: C: Assignment Branch Condition size for api is too high. [27.95/15]
  def api
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:15:3: C: Method has too many lines. [31/10]
  def api
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:22:9: C: Use raise instead of fail to rethrow exceptions.
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Key available"
        ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:22:81: C: Line is too long. [97/80]
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Key available"
                                                                                ^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:30:81: C: Line is too long. [85/80]
        secret = YAML.load_file('/etc/foreman/settings.yaml')[:oauth_consumer_secret]
                                                                                ^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:32:9: C: Use raise instead of fail to rethrow exceptions.
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Secret available"
        ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:32:81: C: Line is too long. [100/80]
        fail "Hostgroup #{resource[:name]} cannot be registered: No OAUTH Consumer Secret available"
                                                                                ^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:36:38: C: Redundant curly braces around a hash parameter.
    @api ||= ApipieBindings::API.new({
                                     ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:37:7: C: Use the new Ruby 1.9 hash syntax.
      :uri => resource[:base_url],
      ^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:37:7: C: Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.
      :uri => resource[:base_url],
      ^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:38:7: C: Use the new Ruby 1.9 hash syntax.
      :api_version => 2,
      ^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:39:7: C: Use the new Ruby 1.9 hash syntax.
      :oauth => {
      ^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:40:9: C: Use the new Ruby 1.9 hash syntax.
        :consumer_key    => key,
        ^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:41:9: C: Use the new Ruby 1.9 hash syntax.
        :consumer_secret => secret
        ^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:43:7: C: Use the new Ruby 1.9 hash syntax.
      :timeout => resource[:timeout],
      ^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:44:7: C: Use the new Ruby 1.9 hash syntax.
      :headers => {
      ^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:45:9: C: Use the new Ruby 1.9 hash syntax.
        :foreman_user => resource[:effective_user],
        ^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:45:51: C: Avoid comma after the last item of a hash.
        :foreman_user => resource[:effective_user],
                                                  ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:47:7: C: Use the new Ruby 1.9 hash syntax.
      :apidoc_cache_base_dir => File.join(Puppet[:vardir], 'apipie_bindings')
      ^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:48:5: C: Indent the right brace the same as the first position after the preceding left parenthesis.
    })
    ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:51:1: C: Extra blank line detected.
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:52:19: W: Unused method argument - value. If it's necessary, use _ or _value as an argument name to indicate that it won't be used. You can also write as smartproxy=(*) if you want the method to accept any arguments but don't care about them.
  def smartproxy=(value)
                  ^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:30: C: Redundant curly braces around a hash parameter.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:32: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                               ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:44: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                           ^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:61: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                                            ^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:81: C: Line is too long. [107/80]
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:100: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                                                                                   ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:105: C: Space inside } missing.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                                                                                        ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:53:106: C: Space inside } missing.
    hostgroups.call(:update, { "id" => id, "hostgroup" => { "puppet_proxy_id" => smartproxy_object["id"]}})
                                                                                                         ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:54:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:63:31: C: Use the new Ruby 1.9 hash syntax.
    smartproxies.call(:index, :search => "name=#{resource[:smartproxy]}")['results'][0]
                              ^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:63:81: C: Line is too long. [87/80]
    smartproxies.call(:index, :search => "name=#{resource[:smartproxy]}")['results'][0]
                                                                                ^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:67:27: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroup ? hostgroup["puppet_proxy_id"] : nil
                          ^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:71:46: C: Use the new Ruby 1.9 hash syntax.
    smartproxy_id ? smartproxies.call(:show, :id => smartproxy_id)["name"] : nil
                                             ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:71:68: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    smartproxy_id ? smartproxies.call(:show, :id => smartproxy_id)["name"] : nil
                                                                   ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:76:1: C: Incorrect indentation detected (column 0 instead of 2).
#
^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:78:68: C: Use the new Ruby 1.9 hash syntax.
    @environment_object = api.resource(:environments).call(:index, :search => "name=#{resource[:environment]}")['results'][0]
                                                                   ^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:78:81: C: Line is too long. [125/80]
    @environment_object = api.resource(:environments).call(:index, :search => "name=#{resource[:environment]}")['results'][0]
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:82:28: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    @hostgroup ? hostgroup["environment_id"] : nil
                           ^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:86:28: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    @hostgroup ? hostgroup["environment_name"] : nil
                           ^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:89:20: W: Unused method argument - value. If it's necessary, use _ or _value as an argument name to indicate that it won't be used. You can also write as environment=(*) if you want the method to accept any arguments but don't care about them.
  def environment=(value)
                   ^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:30: C: Redundant curly braces around a hash parameter.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:32: C: Use the new Ruby 1.9 hash syntax.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                               ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:43: C: Use the new Ruby 1.9 hash syntax.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                          ^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:59: C: Use the new Ruby 1.9 hash syntax.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                                          ^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:81: C: Line is too long. [104/80]
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:97: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                                                                                ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:102: C: Space inside } missing.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                                                                                     ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:90:103: C: Space inside } missing.
    hostgroups.call(:update, { :id => id, :hostgroup => { :environment_id => environment_object["id"]}})
                                                                                                      ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:91:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:100:57: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    @hostgroupclasses = hostgroupclass_api.call(:index, "hostgroup_id" => id)["results"].inject({}){|h,c|
                                                        ^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:100:79: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
    @hostgroupclasses = hostgroupclass_api.call(:index, "hostgroup_id" => id)["results"].inject({}){|h,c|
                                                                              ^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:100:81: C: Line is too long. [105/80]
    @hostgroupclasses = hostgroupclass_api.call(:index, "hostgroup_id" => id)["results"].inject({}){|h,c|
                                                                                ^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:100:100: C: Avoid using {...} for multi-line blocks.
    @hostgroupclasses = hostgroupclass_api.call(:index, "hostgroup_id" => id)["results"].inject({}){|h,c|
                                                                                                   ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:100:103: C: Space missing after comma.
    @hostgroupclasses = hostgroupclass_api.call(:index, "hostgroup_id" => id)["results"].inject({}){|h,c|
                                                                                                      ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:102:25: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      obj ? h.merge(obj["name"] => obj["id"]) : h
                        ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:102:40: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      obj ? h.merge(obj["name"] => obj["id"]) : h
                                       ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:115:33: C: Use the new Ruby 1.9 hash syntax.
    puppetclass_api.call(:show, :id => id)
                                ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:122:3: C: Assignment Branch Condition size for puppetclass= is too high. [20.49/15]
  def puppetclass=(classes)
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:122:3: C: Method has too many lines. [17/10]
  def puppetclass=(classes)
  ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:123:35: C: Avoid using {...} for multi-line blocks.
    class_ids = classes.inject([]){|a,c|
                                  ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:123:38: C: Space missing after comma.
    class_ids = classes.inject([]){|a,c|
                                     ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:125:38: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
        a.push puppetclass_object(c)["id"]
                                     ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:134:16: C: Avoid using {...} for multi-line blocks.
    remove.each{|cid|
               ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:135:41: C: Redundant curly braces around a hash parameter.
      hostgroupclass_api.call(:destroy, {"id" => cid, "hostgroup_id" => id})
                                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:135:41: C: Space inside { missing.
      hostgroupclass_api.call(:destroy, {"id" => cid, "hostgroup_id" => id})
                                        ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:135:42: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      hostgroupclass_api.call(:destroy, {"id" => cid, "hostgroup_id" => id})
                                         ^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:135:55: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      hostgroupclass_api.call(:destroy, {"id" => cid, "hostgroup_id" => id})
                                                      ^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:135:75: C: Space inside } missing.
      hostgroupclass_api.call(:destroy, {"id" => cid, "hostgroup_id" => id})
                                                                          ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:138:13: C: Avoid using {...} for multi-line blocks.
    add.each{|cid|
            ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:139:40: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      hostgroupclass_api.call(:create, "puppetclass_id" => cid, "hostgroup_id" => id)
                                       ^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:139:65: C: Prefer single-quoted strings when you don't need string interpolation or special symbols.
      hostgroupclass_api.call(:create, "puppetclass_id" => cid, "hostgroup_id" => id)
                                                                ^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:139:81: C: Line is too long. [85/80]
      hostgroupclass_api.call(:create, "puppetclass_id" => cid, "hostgroup_id" => id)
                                                                                ^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:142:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:152:42: C: Use the new Ruby 1.9 hash syntax.
    @hostgroup = hostgroups.call(:index, :search => "name=#{resource[:name]}")['results'][0]    # end
                                         ^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:152:81: C: Line is too long. [101/80]
    @hostgroup = hostgroups.call(:index, :search => "name=#{resource[:name]}")['results'][0]    # end
                                                                                ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:152:93: C: Unnecessary spacing detected.
    @hostgroup = hostgroups.call(:index, :search => "name=#{resource[:name]}")['results'][0]    # end
                                                                                            ^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:153:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:162:5: C: Do not leave space between ! and its argument.
    ! id.nil?
    ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:166:30: C: Redundant curly braces around a hash parameter.
    hostgroups.call(:create, {
                             ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:167:7: C: Use the new Ruby 1.9 hash syntax.
      :hostgroup => {
      ^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:167:7: C: Use 2 spaces for indentation in a hash, relative to the first position after the preceding left parenthesis.
      :hostgroup => {
      ^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:168:9: C: Use the new Ruby 1.9 hash syntax.
        :name => resource[:name]
        ^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:170:5: C: Indent the right brace the same as the first position after the preceding left parenthesis.
    })
    ^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:171:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:176:31: C: Use the new Ruby 1.9 hash syntax.
    hostgroups.call(:destroy, :id => id)
                              ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:178:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:183:23: C: Use the new Ruby 1.9 hash syntax.
    hostgroup.refresh(:id => id)
                      ^^^^^^
lib/puppet/provider/foreman_hostgroup/rest_v2.rb:184:3: W: Avoid rescuing the Exception class. Perhaps you meant to rescue StandardError?
  rescue Exception => e
  ^^^^^^^^^^^^^^^^^^^^^

1 file inspected, 94 offenses detected
```
I also have to say there's no .rubocop.yml configured, so this is not using the current Foreman core ruleset, but only the rubocop defaults. i.e. Ruby 1.9 syntax is disabled for Foreman.

flyinbutrs commented:
Ok, I'll give that a shot...  I would point out that some of those complaints are asking for the ruby 1.9 hash syntax which I removed to try to pass the travis-ci tests for 1.8 :)

mmoll commented:
right, we still have to support 1.8 there, so that output is dangerous to follow anyway

mmoll commented:
what is the smartproxy parameter supposed to take as argument? this should set the puppet master for the host group, correct?

flyinbutrs commented:
It's expecting the "name" value from the smartproxy, and sets the puppetmaster and puppet_ca.  I can add more granular options if needed, this met my needs, so that's what I did :)

mmoll commented:
ah, I see, after a hard reload of that page I see the values... I think seperate options for CA and master would make sense and if possible also Organization and Location. Otherwise it's working fine here.

flyinbutrs commented:
Ok, I will add them.

flyinbutrs commented:
Hey, so I'm just starting to look at this, and noticed a few things...

1) I don't actually set the CA, I'm just setting the "puppet_proxy_id" on the hostgroup, and it's inferring the CA. So maybe add a separate field for CA and leave the current "smartproxy" property? Or should I replace "smartproxy" with two fields "puppetmaster" and "puppet_ca"? Or just add the other two in addition to the "smartproxy"?

2) I don't see any "Organization" or "Location" options in the gui anywhere...  where are they located? Also, I see them as resources in the API, though not as options on the hostgroups resource in the API.

flyinbutrs commented:
@mmoll see above for other questions, but in the meantime, I have modified as follows:

Removed: smartproxy
Added: puppetmaster
Added: puppet_ca
Added: parent

Still unsure about the org / loc

mmoll commented:
Orgs/Locs have to be enabled specially, if one wants to use these. In my setup I do use locations, so I have to tag all the resources with locations to be useful. It's also in the docs (the last two options): http://theforeman.org/api/1.9/apidoc/v2/hostgroups/update.html

flyinbutrs commented:
Ok, added support for orgs and locations...  aaaaand literally just realized I used locations to update instead of location_ids.  Both seem to work, not sure if it matters.  locations appears not to be listed in the API docs which makes me wary.

mmoll commented:
what do you mean with "Locations appears not to be listed in the API docs"? What docs are you using specifically?

flyinbutrs commented:
According to the link you send, I should update using "location_ids" as an array of IDs.  I am currently updating as follows:

```ruby
hostgroups.call(:update, { :id => id, :hostgroup => { :locations => new_locations }})
```

And new_locations is an array of the return of:

```ruby
location_api.call(:index, :search => "name=#{name}")['results'][0]
```

mmoll commented:
Hm, maybe a consequence of locations beeing able to be nested, but if you think the API is inconsistent, please open a redmine issue.

mmoll commented:
One little thing, when the locations are updated, the output is rendering all locations into one string, wh
ich is confusing:
`Notice: /Stage[main]/Main/Foreman_hostgroup[provtest]/locations: locations changed [] to 'Home GCSC2'` other then that this seems to work fine for me. :+1:

flyinbutrs commented:
That's definitely odd, but I'm not really sure why it's doing that.  The output of "locations" should be an array of names.

flyinbutrs commented:
So do I need anything else before this gets merged?

mmoll commented:
@domcleal @mbacovsky could you have a look?

domcleal commented:
> According to the link you send, I should update using "location_ids" as an array of IDs. I am currently updating as follows:
> 
> hostgroups.call(:update, { :id => id, :hostgroup => { :locations => new_locations }})

Yeah, using :location_ids or :location_names would be the correct method.  I'm not sure that :locations will ever work consistently, avoid it.

flyinbutrs commented:
Didn't realize there were location_names and organization_names options...  that simplifies my code a lot, so I've switched to that format.

Also, found a bug :) (http://projects.theforeman.org/issues/12263)

flyinbutrs commented:
A vaguely related question...  we use this plugin heavily:

https://github.com/theforeman/foreman_default_hostgroup

One thing I will be doing in one place or another, is setting it up so that the default_hostgroups.yaml file is concat generated so you can define it as part of the manifest along with the hostgroups.  The question is, should that be a part of this module, or does it have to be elsewhere because it's a plugin?

mmoll commented:
Have a look onto how plugins are configured in this module, the best example is probably discovery or openscap.

`::foreman::plugin::default_hostgroup` at the moment is only installing the package for the plugin but this can get extended to generate its config file with parameters, which would be a nice change, but to be handled by a different PR.

flyinbutrs commented:
I will take a look, and see if I can figure out how to setup the plugin.. In the meantime, I've found a few bugs and am still chasing down one more.  I'll take another stab tomorrow, going to bed for now.  Right now, Puppetclasses will constantly get set if a hostgroup is assigned a class that it's inherited from a parent because I wasn't recursively puppet class assignments for all parent hostgroups.

flyinbutrs commented:
Ok, bugs squashed.

flyinbutrs commented:
I found a couple more edge cases after putting it into use in some test environments, now also patched.

flyinbutrs commented:
any movement on this?

mmoll commented:
I'm in favour of merging this, just want the API people have a look on it.

flyinbutrs commented:
No problem, just wanted to make sure it wasn't forgotten :)

Works out, because I keep finding little bugs to squash anyway.

ekohl commented on lib/puppet/provider/foreman_hostgroup/rest_v2.rb:
```diff
@@ -0,0 +1,316 @@
+    super + 1
+  end
+
+  def settings
```
It doesn't have to be done in this PR, but we should keep this part consistent for all apipie resources. Currently that's only the smartproxy, so shouldn't be much work. I do wonder if we can abstract this to some helper class.

ekohl commented:
I can't review the ruby code too well, but it looks good to me.

flyinbutrs commented:
rebased on the latest version, now travis is fully passing. I've been using this in production for several months...  any chance of a merge? I can squash the commits first if that's desired

mmoll commented:
@flyinbutrs yes, please squash all the commits together, I think this is good to go :carousel_horse:

flyinbutrs commented:
Ok, commit squashed. I'm going to also add the autoload of section that @ekohl recommended to the smartproxy provider in a separate merge

mmoll commented:
The report processor test error on Ruby 1.8 looks unrelated, anybody an idea about that?

flyinbutrs commented:
no idea, it happened on my other PR as well (#394)

mmoll commented:
yeah, it's reproducible unrelated. merged, thanks @flyinbutrs!
